File: 0239-sliding-window-maximum/0239-sliding-window-maximum.py
from collections import deque
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        queue = deque()
        ans = []
        n = len(nums)
        for i in range(k):
            while len(queue) > 0 and queue[-1] < nums[i]:
                queue.popleft()
            queue.append(nums[i])
        for i in range(k, n):
            ans.append(queue[0])
            if nums[i - k] == queue[0]:
                queue.popleft()
            while len(queue) > 0 and queue[-1] < nums[i]:
                queue.pop()
            queue.append(nums[i])
        ans.append(queue[0])
        print(i, queue)
        return ans

File: 0239-sliding-window-maximum/0239-sliding-window-maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        deque<int>que;
        vector<int>ans;
        for(int i=0;i<k;i++){
            while(!que.empty() && que.back()<nums[i]){
                que.pop_back();
            }
            que.push_back(nums[i]);
        }
        ans.push_back(que.front());
        for(int i=k;i<nums.size();i++){
            while(!que.empty() && que.back()<nums[i]){
                que.pop_back();
            }
            que.push_back(nums[i]);
            if(que.front()==nums[i-k]){
                que.pop_front();
            }
            ans.push_back(que.front());
        }
        return ans;
    }
};